File: sdk/packages/node/iii/src/triggers.ts
/**
 * Configuration passed to a trigger handler when a trigger instance is
 * registered or unregistered.
 *
 * @typeParam TConfig - Type of the trigger-specific configuration.
 */
export type TriggerConfig<TConfig> = {
  /** Trigger instance ID. */
  id: string
  /** Function to invoke when the trigger fires. */
  function_id: string
  /** Trigger-specific configuration. */
  config: TConfig
  /** Arbitrary metadata attached to the trigger. */
  metadata?: Record<string, unknown>
}

/**
 * Handler interface for custom trigger types. Passed to
 * `IIIClient.registerTriggerType`.
 *
 * @typeParam TConfig - Type of the trigger-specific configuration.
 *
 * @example
 * ```typescript
 * const handler: TriggerHandler<{ interval: number }> = {
 *   async registerTrigger({ id, function_id, config }) {
 *     // Set up periodic invocation
 *   },
 *   async unregisterTrigger({ id, function_id, config }) {
 *     // Clean up
 *   },
 * }
 * ```
 */
export type TriggerHandler<TConfig> = {
  /** Called when a trigger instance is registered. */
  registerTrigger(config: TriggerConfig<TConfig>): Promise<void>
  /** Called when a trigger instance is unregistered. */
  unregisterTrigger(config: TriggerConfig<TConfig>): Promise<void>
}

File: sdk/packages/rust/iii/src/triggers.rs
use std::sync::Arc;

use async_trait::async_trait;
use serde_json::Value;

use crate::error::Error;

/// Configuration passed to a [`TriggerHandler`] when a trigger instance is
/// registered or unregistered.
#[derive(Debug, Clone)]
pub struct TriggerConfig {
    /// Trigger instance ID.
    pub id: String,
    /// Function to invoke when the trigger fires.
    pub function_id: String,
    /// Trigger-specific configuration.
    pub config: Value,
    /// Arbitrary metadata attached to the trigger.
    pub metadata: Option<Value>,
}

/// Handler trait for custom trigger types. Implement this and pass to
/// [`IIIClient::register_trigger_type`](crate::IIIClient::register_trigger_type).
#[async_trait]
pub trait TriggerHandler: Send + Sync {
    /// Called when a trigger instance is registered.
    async fn register_trigger(&self, config: TriggerConfig) -> Result<(), Error>;
    /// Called when a trigger instance is unregistered.
    async fn unregister_trigger(&self, config: TriggerConfig) -> Result<(), Error>;
}

/// Handle returned by [`IIIClient::register_trigger`](crate::IIIClient::register_trigger).
/// Call [`unregister`](Trigger::unregister) to remove the trigger from the engine.
#[derive(Clone)]
pub struct Trigger {
    unregister_fn: Arc<dyn Fn() + Send + Sync>,
}

impl Trigger {
    pub fn new(unregister_fn: Arc<dyn Fn() + Send + Sync>) -> Self {
        Self { unregister_fn }
    }

    /// Remove this trigger from the engine.
    pub fn unregister(&self) {
        (self.unregister_fn)();
    }
}

#[cfg(test)]
mod tests {
    use std::sync::{
        Arc,
        atomic::{AtomicBool, Ordering},
    };

    use super::*;

    #[test]
    fn trigger_unregister_calls_closure() {
        let called = Arc::new(AtomicBool::new(false));
        let called_ref = called.clone();
        let trigger = Trigger::new(Arc::new(move || {
            called_ref.store(true, Ordering::SeqCst);
        }));

        trigger.unregister();

        assert!(called.load(Ordering::SeqCst));
    }
}
